docs/syntax/syntax.md

```diff
@@ -205,95 +205,429 @@ Is below, but it won't be parsed into the document.
 
 (syntax/referencing)=
 
-## Markdown Links and Referencing
+## Links and Referencing
 
-Markdown links are of the form: `[text](link)`.
+CommonMark links come in three forms ([see the spec](https://spec.commonmark.org/0.30/#links)):
 
-If you set the configuration `myst_all_links_external = True` (`False` by default),
-then all links will be treated simply as "external" links.
-For example, in HTML outputs, `[text](link)` will be rendered as `<a href="link">text</a>`.
+*Autolinks* are [URIs][uri] surrounded by `<` and `>`:
 
-Otherwise, links will only be treated as "external" links if they are prefixed with a scheme,
-configured with `myst_url_schemes` (by default, `http`, `https`, `ftp`, or `mailto`).
-For example, `[example.com](https://example.com)` becomes [example.com](https://example.com).
+```md
+<scheme:path?query#fragment>
+```
 
-:::{note}
-The `text` will be parsed as nested Markdown, for example `[here's some *emphasised text*](https://example.com)` will be parsed as [here's some *emphasised text*](https://example.com).
+*Inline links* allow for optional explicit text and titles (in HTML titles are rendered as tooltips):
+
+```md
+[Explicit *Markdown* text](destination "optional explicit title")
+```
+
+or, if the destination contains spaces,
+
+```md
+[text](<a destination>)
+```
+
+*Reference links* define the destination separately in the document, and can be used multiple times:
+
+```md
+[Explicit *Markdown* text][label]
+
+[label]: destination "optional explicit title"
+```
+
+MyST, supports the following destination types:
+
+|     Link Type      |            Auto            |          Inline           | Single Page[^sp] |
+| :----------------- | :------------------------- | :------------------------ | :--------------: |
+| External URL       | `<https://example.com>`    | `[](https://example.com)` |        ✅         |
+| Local file path    | `<path:file.txt>`          | `[](file.txt)`            |        ❌         |
+| Project document   | `<project:file.md>`        | `[](file.md)`             |        ❌         |
+| Local target       | `<project:.#target>`       | `[](.#target)`            |        ✅         |
+| Target in document | `<project:target.md#file>` | `[](file.md#target)`      |        ❌         |
+| Target in project  | `<project:#target>`        | `[](#target)`             |     ~[^hash]     |
+| Cross-project      | `<myst:key#target>`        | `[](myst:key#target)`     |        ❌         |
```

Comment:
Following from the above comment, the `myst` syntax as used here confuses me a bit, because here we're explicitly referring to a target in another project, but now we _don't_ use the `project` scheme, we use a `myst` scheme. To me, the syntax `<project:key#target>`  reads like "reference `target` in project `key`". I'm not sure how to similarly interpret the `myst:` scheme.

Reply:
well @rowanc1 wanted to use `myst:` here, so I'll let you take it up with him 😉 

I think the logic for handling internal references and external inventory references is perhaps different enough though to warrant a different scheme

Reply:
I took another user through this and they also were confused by the syntax.

We could potentially merge what is currently `project:` and `myst:` into a single `project:`, and if the key isn't given we only look to the local project? That would still keep the logic separate.

I liked `myst:` because it was short, reinforced the branding of the project and the concept could be extended to other, richer, content links in the future. I am not super tied to it, and I do think that `<project:jupyterbook#getting-started>` is maybe much clearer?

Reply:
I'm fine with calling it `myst:` if we can have a clear and easily memorable rationale for what the naming means and what behavior it should evoke. The confusing thing to me is that `[](myst:`, `[](file:)`, and `[](project:)` all mean different kinds of things. `file:` is the thing `file:` refers to, `project:` is "a thing within the thing specified after `project:` and `myst:` is "anything that has something to do with MyST".

If we used `myst:` it feels more natural to me to use it for *everything* and define the first word after as the way to interpret what comes next. e.g. `[](myst:target?project=executablebooks#mytarget)`, `[](myst:file#hello.txt`, `[](myst:someotherthing?foo=bar)`

Reply:
> if the key isn't given we only look to the local project?

what if you want to search through all external projects?

Reply:
> `file:` is the thing `file:` refers to

I'd note that unfortunately we cannot use the `file` schema easily, because markdown-it disallows it, on grounds of security: https://github.com/markdown-it/markdown-it/blob/08444a5c1c84440f0c03a23c26d5cf57175e7575/lib/index.js#L32

Reply:
Orginally I had `myst:` as `inv:` because that is what t is actually referring to; a sphinx `objects.inv` file format, if we are not using `myst` hen I think it should be that, to clarify exactly what it is referring to

Reply:
I would actually then consider changing `project` to `myst`, because it is referring to a target specifically identified by MyST

Reply:
It's reasonable to me to use `myst:` as a catch-all schema meaning "this will refer to an object that is known to a MyST system". Could we then just use that for everything rather than defining other schemas that are specific to the referenced object? What if we made three assumptions:

- All contents of a markdown link are the `myst` scheme, unless they start with a web-based scheme like `http` etc.
- All myst schemes have a `kind` parameter that defines the rules that will be used to resolve the reference. It follows this logic:
	- If the path is a target in the MyST system, kind is `target` unless explicitly specified
	- If the path is not a target in the MyST system but is a valid filepath, kind is `file` unless explicitly specified
	- Other `kind`s could be registered, potentially along with a validation rule to auto-detect a given "kind" if neither `target` nor `file` were triggered.
- All myst schemes have a `project` parameter to optionally restrict the search to a specific project inventory. If not given, all project inventories are searched, beginning with the local inventory, and the first match is returned.

So for example:

- To reference a target: `[](#mytarget)` which implies `[](myst:?kind=target#mytarget)`
    - To restrict the search to a specific project: `[](?project=myproj#mytarget)`
- To reference a target that is a document: `[](mydoc.md)` which implies `[](myst:mydoc.md?kind=target)`
- To reference a target that is _inside_ a document: `[](mydoc.md#mytarget)` which implies `[](myst:mydoc.md?kind=target#mytarget)`
- To download files directly: `[](folder/path.txt)` which implies `[](myst:folder/path.txt?kind=file)` because `folder/path.txt` is not in the MyST system's list of targets, but is a valid path.
- To raise a warning: `[](folder/nonexistentfile.foo)` which doesn't exist in any myst inventory, and isn't a local file, and doesn't have some other `kind` registered to make sense of it.

Reply:
@choldgraf does this imply that `[](#mytarget)` should search through all intersphinx inventories, as well as the local project?

Reply:
> All contents of a markdown link are the myst scheme, unless they start with a web-based scheme like http etc.

No. Every link should have at least a qualifying prefix; either `#` or a `schema:`, otherwise it makes it very brittle to understand what the destination is referring to.

Reply:
> does this imply that [](#mytarget) should search through all intersphinx inventories, as well as the local project?

I think the logic would be something like "search through all loaded inventories, starting with the local one, and return the first target that matches"

> No. Every link should have at least a qualifying prefix; either # or a schema:, otherwise it makes it very brittle to understand what the destination is referring to.

I suggested assuming `myst:` schema because I thought that was the only way to let it resolve something like: `[](mydoc.md)`. Are you proposing that users should instead have to write `[](myst:mydoc.md)` if they want to link to a local document?

Reply:
> To reference a Python class w/ a kind that was registered by autodoc's presence: `[](MyClass)` which implies `[](myst:MyClass?kind=autodoc)`

there is no way in sphinx to know if it was registered by autodoc, this is also not something that can go in to a MyST spec

> return the first target that matches

I'm very against it just matching the first thing, without any warning. This is basically going to introduce exactly the problem that sphinx had, which led to them introduce the `external` role (https://github.com/sphinx-doc/sphinx/issues/9682)

If the "shorthand" matches anything, anyone with a reasonably complex project will not be able to use it, and will then have to use a horribly complex query string, which will just make them go back to using roles.

Reply:
> Are you proposing that users should instead have to write [](myst:mydoc.md) if they want to link to a local document?

No. `[](mydoc.md)` is allowed, because Commonmark editors/previewers like VS Code and GitHub, but if you also say that `[](mydoc.md)` could also potentially signify that `mydoc.md` is a local/project/intersphinx target, then you have to introduce loads of complex logic into every implementation of MyST.

Similarly for previewers/LSPs, if you know explicitly that it is a reference to an external inventory, you can easily pull up the reference to it, if not then you can't

Reply:
So then the logic is:

- If there's a markdown link without `myst:`, then check if it is a valid local path.
- If the path exists, check if it is in the myst document store. If so, treat it as `[](myst:myfile.md)`
- If exists but isn't in the document store, then treat it as `[](myst:myfile.md?kind=file)`
- If the path doesn't exist, raise a warning.
- If the link starts with `#` then assume that it is a MyST target.

So how would you reference the `MyClass` autodoc class above?

Reply:
> So how would you reference the MyClass autodoc class above?

autodoc references are not special-cases in any way, they just create a target, wherever the `MyClass` is documented, and a reference name to it being the fully qualified name (e.g. `module.submodule.MyClass`).

If you want to reference it you do the same as any other target `[](#module.submodule.MyClass)` or, with the logic I've put in, you can use `[](#*MyClass)` to signify a simple regex match of any target ending in `MyClass`.

Also as with any target, you can scope by domain and object type, in this case `py:class`

Reply:
> So then is this the logic?

Nearly:

- If there's a markdown link without a schema, then check if it is a valid local path.
- If the path exists, check if it is in the myst document store. If so, treat it as `[](myst:myfile.md)`
- If exists but isn't in the document store, then treat it as a raw download `[](myst:myfile.md?kind=file)`
- If the link starts with `#` (or `.#`/`? `/`.?`) then assume that it is a MyST target.
- If the link starts with a `schema:` get that and try to match it:
  - if it's in the "external list" (`http`, `https` etc) treat it as external
  - if it's known to MyST (e.g. `myst`) use that logic
  - Potentially in the future allow plugins to extend to handling other schemes
- If there is still no match, then emit a warning.
  - Currently then, mainly for back-compatibility considerations, `[](unknown)` is treated as if it was actually `[](#unknown)`, and progresses along that logic

Reply:
the basic thing is if you have a `name` and you say it can relate to any sphinx/intersphinx target, then you have potentially four variables to filter by:

- the project of the target, i.e. a key in the intersphinx mapping
- the docname of the target
  - this information is not specifically available from objects.inv, only for local targets
  - end event then, sphinx actually overwrites any conflicting targets (in the same domain/object type group) from different documents, so filtering by document does not work that well (except for MyST anchors, which are a different thing)
- the domain of the target
- the object type of the target

You need some way to specify those filters in a nice syntax

Reply:
So you could say that for `[](myst:xxx#name)`, `xxx` is **either** a local document path or key to an `objects.inv` path/URL.

You then would have to make sure obviously that no key conflicts with a file path, but this should not be too difficult, as long as your key does not end in `.md` or something.

I think `[](#name)` should always search through everything both locally and `object.inv`, warn if there is multiple matches, then can e.g. choose the match based on some rules like local first.

You don't have a syntax in this way to say only search in `objects.inv`'s, and not locally, for this you could do something like `[](*#name)` 🤷 

You then need to provide a filter for domain/objects.
As I've mentioned already, the most terse format would be `[](?domain:object#name)`, and if you want all domains use `[](?*:object#name)`.
Less terse, but keeping with the "usual" query format, would be `[](?f=domain:object#name)`

Reply:
Note for the above, links to "unknown" local files (aka ones to be downloaded) is not part of this, I don't think that should be part of the `myst` scheme as its quite different

Also `[](myst:key)` could either be meaningless, i.e. you just emit a warning, or you do something like create a link to the "base" URL